crypto/cms/cms_sd.c

```diff
@@ -327,8 +359,15 @@ CMS_SignerInfo *CMS_add1_signer(CMS_ContentInfo *cms,
         }
     }
 
-    if (!(flags & CMS_KEY_PARAM) && !cms_sd_asn1_ctrl(si, 0))
-        goto err;
+    if (!(flags & CMS_KEY_PARAM)) {
+        EVP_MD_CTX *tmp_mctx = EVP_MD_CTX_new();
+        int ok = (EVP_DigestSignInit(tmp_mctx, NULL, md, NULL, pk) > 0);
+
+        ok = ok && cms_sd_asn1_ctrl(si, EVP_MD_CTX_pkey_ctx(tmp_mctx), 0);
+        EVP_MD_CTX_free(tmp_mctx);
+        if (!ok)
+            goto err;
```

Comment:
Should we raise a specific error for this case?

Reply:
This is a situation where we usually count on the called function to use that already...